### ISSUE TITLE
Reptile: A Scalable Meta-Learning Algorithm

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#354]
https://blog.openai.com/reptile/

**Comment by sweaterr** [comment]
We’ve developed a simple meta-learning algorithm called Reptile which works by repeatedly sampling a task, performing stochastic gradient descent on it, and updating the initial parameters towards the final parameters learned on that task.

Meta-learning is the process of learning how to learn.

**Comment by sweaterr** [comment]
## How Reptile Works

Like MAML, Reptile seeks an initialization for the parameters of a neural network, such that the network can be fine-tuned using a small amount of data from a new task.
MAML과 마찬가지로 Reptile은 신경망의 매개 변수에 대한 초기화를 시도하여 네트워크가 새로운 작업의 소량의 데이터를 사용하여 미세 조정될 수 있도록합니다.

But while MAML unrolls and differentiates through the computation graph of the gradient descent algorithm, Reptile simply performs stochastic gradient descent (SGD) on each task in a standard way — it does not unroll a computation graph or calculate any second derivatives.
그러나 MAML이 그라디언트 디센트 알고리즘의 계산 그래프를 전개하고 구별하는 동안 Reptile은 표준 그래프에서 각 작업에 대한 확률 적 그라디언트 디센트 (SGD)를 단순히 수행합니다. 계산 그래프를 풀거나 두 번째 파생 값을 계산하지 않습니다.

This makes Reptile take less computation and memory than MAML.
이것은 파충류가 MAML보다 적은 계산과 메모리를 사용하도록 만든다.

The pseudocode is as follows:.
의사 코드는 다음과 같습니다.

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/39028187-592766d2-4490-11e8-86dc-c061d9c143f0.png)

**Comment by sweaterr** [comment]
As an alternative to the last step, we can treat  Φ − W  as a gradient and plug it into a more sophisticated optimizer like Adam.
마지막 단계의 대안으로 Φ - W를 그라디언트로 취급하고 Adam과 같은보다 정교한 옵티 마이저에 플러그인 할 수 있습니다.

**Comment by sweaterr** [comment]
It is at first surprising that this method works at all.
처음에는이 방법이 전혀 효과가 없다는 것이 놀랍습니다.

If  k = 1 , this algorithm would correspond to “joint training” — performing SGD on the mixture of all tasks.
k = 1 인 경우이 알고리즘은 모든 작업을 혼합하여 SGD를 수행하는 "공동 학습"에 해당합니다.

While joint training can learn a useful initialization in some cases, it learns very little when zero-shot learning is not possible (e.g, when the output labels are randomly permuted).
공동 훈련은 경우에 따라 유용한 초기화를 배울 수 있지만, 영점 조정 학습이 불가능할 때 (예 : 출력 레이블이 무작위로 변경되는 경우) 거의 학습하지 않습니다.

Reptile requires  k > 1 , where the update depends on the higher-order derivatives of the loss function; as we show in the paper, this behaves very differently from k=1 (joint training).
파충류는 k> 1을 필요로하는데, 여기서 갱신은 손실 함수의 고차 미분에 의존한다. 우리가 논문에서 보여 주듯이 이것은 k = 1 (공동 훈련)과 매우 다르게 행동합니다

**Comment by sweaterr** [comment]
To analyze why Reptile works, we approximate the update using a Taylor series.
Reptile이 작동하는 이유를 분석하기 위해 Taylor 시리즈를 사용하여 업데이트를 근사합니다.

We show that the Reptile update maximizes the inner product between gradients of different minibatches from the same task, corresponding to improved generalization.
파충류 업데이트가 개선 된 일반화에 해당하는 동일한 작업에서 다른 미니 바 그라디언트 사이의 내부 제품을 최대화한다는 것을 보여줍니다.

This finding may have implications outside of the meta-learning setting for explaining the generalization properties of SGD.
이 결과는 SGD의 일반화 속성을 설명하기위한 메타 학습 설정 이외의 사항에 영향을 미칠 수 있습니다.

Our analysis suggests that Reptile and MAML perform a very similar update, including the same two terms with different weights.
우리의 분석은 Reptile과 MAML이 서로 다른 가중치를 갖는 동일한 두 용어를 포함하여 매우 유사한 업데이트를 수행함을 제시합니다.

**Comment by sweaterr** [comment]
In our experiments, we show that Reptile and MAML yield similar performance on the Omniglot and Mini-ImageNet benchmarks for few-shot classification.
우리의 실험에서, 우리는 파충류와 MAML이 Omniglot과 Mini-ImageNet 벤치 마크에서 유사한 결과를 거의 촬영하지 못한다는 것을 보여줍니다.

Reptile also converges to the solution faster, since the update has lower variance.
파충류는 또한 업데이트가 더 낮은 분산을 가지기 때문에 솔루션에 더 빨리 수렴합니다.

**Comment by sweaterr** [comment]
Our analysis of Reptile suggests a plethora of different algorithms that we can obtain using different combinations of the SGD gradients.
파충류에 대한 우리의 분석은 SGD 그라디언트의 다양한 조합을 사용하여 얻을 수있는 수많은 알고리즘을 제안합니다.

In the figure below, assume that we perform k steps of SGD on each task using different minibatches, yielding gradients  g 1 , g 2 , … , g k.
아래 그림에서 그라디언트 g 1, g 2, ..., g k를 산출하는 다양한 minibatches를 사용하여 각 작업에 대해 k 단계의 SGD를 수행한다고 가정합니다.

The figure below shows the learning curves on Omniglot obtained by using each sum as the meta-gradient.
아래 그림은 각 합계를 메타 그래디언트로 사용하여 얻은 Omniglot의 학습 곡선을 보여줍니다.

g 2  corresponds to first-order MAML, an algorithm proposed in the original MAML paper.
g 2는 1 차 MAML에 해당하며, 원래 MAML 논문에서 제안 된 알고리즘입니다.

Including more gradients yields faster learning, due to variance reduction.
더 많은 그라디언트를 포함 시키면 분산 감소로 인해 학습 속도가 빨라집니다.

Note that simply using  g 1 (which corresponds to  k = 1 ) yields no progress as predicted for this task since zero-shot performance cannot be improved.
제로 샷 성능을 향상시킬 수 없기 때문에 g 1 (k = 1에 해당)을 사용하면이 작업에 대해 예상대로 진행되지 않습니다.

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/39029554-5db5a68a-4497-11e8-89c0-3be6e212a17a.png)
